File: src/ip_to_cidr_751/solution.py
from typing import List


class Solution:
    def ip_to_int(self, ip: str):
        result = 0

        for part in ip.split("."):
            octet = int(part)
            result = (result << 8) + octet

        return result

    def trailing_zeros(self, num: int):
        if num == 0:
            return 32

        zeros = 0
        while num & 1 == 0:
            zeros += 1
            num >>= 1

        return zeros

    def to_cidr(self, ip: int, bits: int) -> str:
        print(bits)
        return f"{self.to_ip(ip)}/{bits}"

    def to_ip(self, ip: int) -> str:
        octets = []
        for i in range(3, -1, -1):
            octet = (ip >> (i * 8)) & 0xFF
            octets.append(str(octet))
        return ".".join(octets)

    def ipToCIDR(self, ip: str, n: int) -> List[str]:
        cidrs = []
        num = self.ip_to_int(ip)
        while n > 0:
            zeros = self.trailing_zeros(num)

            print(zeros)
            while (1 << zeros) > n:
                zeros -= 1

            cnt = 1 << zeros
            print(zeros)
            cidrs.append(self.to_cidr(num, 32 - zeros))
            num += cnt
            n -= cnt

        return cidrs


def test():
    solution = Solution()
    print(solution.ipToCIDR("255.0.0.7", 10))
    print(solution.ipToCIDR("117.145.102.62", 8))
    print(solution.ipToCIDR("0.0.0.0", 1))

File: src/ip_to_cidr_751/solution.ts
function ipToCIDR(ip: string, n: number): string[] {
  const ipToInt = (ip: string): number => {
    let result = 0;
    for (const part of ip.split(".")) {
      const octet = parseInt(part, 10);
      result = (result << 8) + octet;
    }
    return result;
  };

  const trailingZeros = (num: number): number => {
    if (num === 0) {
      return 32;
    }

    let zeros = 0;
    while ((num & 1) === 0) {
      zeros++;
      num >>= 1;
    }
    return zeros;
  };

  const toCIDR = (ip: number, bits: number): string => {
    return `${toIp(ip)}/${bits}`;
  };

  const toIp = (ip: number): string => {
    const octets: string[] = [];
    for (let i = 3; i >= 0; i--) {
      const octet = (ip >> (i * 8)) & 0xff;
      octets.push(octet.toString());
    }
    return octets.join(".");
  };

  const powerOfTwo = (exponent: number): number => {
    return Math.pow(2, exponent);
  };

  const cidrs: string[] = [];
  let num = ipToInt(ip);

  while (n > 0) {
    let zeros = trailingZeros(num);

    while (powerOfTwo(zeros) > n) {
      zeros--;
    }

    const cnt = 1 << zeros;
    cidrs.push(toCIDR(num, 32 - zeros));
    num += cnt;
    n -= cnt;
  }

  return cidrs;
}

export function test(): void {
  console.log(ipToCIDR("255.0.0.7", 10));
  console.log(ipToCIDR("117.145.102.62", 8));
  console.log(ipToCIDR("0.0.0.0", 1));
}
